packages/compose/src/higher-order/with-instance-id/index.tsx

```diff
@@ -13,15 +13,18 @@ import useInstanceId from '../../hooks/use-instance-id';
  * A Higher Order Component used to be provide a unique instance ID by
  * component.
  */
-const withInstanceId: < Inner extends ComponentType< any > >(
-	Inner: Inner
-) => ComponentType<
-	Omit< ComponentProps< Inner >, 'instanceId' >
-> = createHigherOrderComponent( ( WrappedComponent ) => {
-	return ( props ) => {
-		const instanceId = useInstanceId( WrappedComponent );
-		return <WrappedComponent { ...props } instanceId={ instanceId } />;
-	};
-}, 'instanceId' );
+const withInstanceId = createHigherOrderComponent<
+	< Props extends ComponentProps< any > >(
+		Inner: ComponentType< Props >
+	) => ComponentType< Props & { instanceId: string | number } >
+>(
+	( WrappedComponent ) => ( props ) => (
+		<WrappedComponent
+			{ ...props }
+			instanceId={ useInstanceId( WrappedComponent ) }
+		/>
+	),
+	'instanceId'
+);
```

Comment:
There are several things wrong about the `withInstanceId` types:
- `extends ComponentProps< any >` is redudant because `ComponentProps` never gets a reasonable component type to infer props from. The result is either `{}` or `any`.
- the `withInstanceId` HOC accepts also components that don't accept `instanceId` as a prop, which is an undetected type error.
- the `withInstanceId` HOC returns a component that accepts the `instanceId` prop, which is an error, too. The outer component doesn't accept that prop, because it's supposed to be injected.